DigitalLearningSolutions.Web.Tests/Controllers/Register/ClaimAccountControllerTests.cs

```diff
@@ -226,9 +255,332 @@ public void
+                .WithActionName("AccessDenied");
+        }
+
+        [Test]
+        public void LinkDlsAccountGet_with_no_existing_user_redirects_to_AccessDenied()
+        {
+            // Given
+            A.CallTo(
+                () => userDataService.GetUserIdAndCentreForCentreEmailRegistrationConfirmationHashPair(
+                    DefaultEmail,
+                    DefaultCode
+                )
+            ).Returns((null, null, null));
+
+            // When
+            var result = controllerWithLoggedInUser.LinkDlsAccount(DefaultEmail, DefaultCode);
+
+            // Then
+            A.CallTo(
+                () => userDataService.GetUserIdAndCentreForCentreEmailRegistrationConfirmationHashPair(
+                    DefaultEmail,
+                    DefaultCode
+                )
+            ).MustHaveHappenedOnceExactly();
+
+            result.Should().BeRedirectToActionResult().WithControllerName("LearningSolutions")
+                .WithActionName("AccessDenied");
+        }
+
+        [Test]
+        public void
+            LinkDlsAccountGet_when_the_logged_in_user_already_has_a_delegate_account_at_the_centre_redirects_to_AccountAlreadyExists()
+        {
+            // Given
+            var expectedModel = GetViewModel();
+            var userEntity = new UserEntity(
+                UserTestHelper.GetDefaultUserAccount(),
+                new List<AdminAccount>(),
+                new List<DelegateAccount> { UserTestHelper.GetDefaultDelegateAccount(centreId: DefaultCentreId) }
+            );
+
+            A.CallTo(
+                () => userDataService.GetUserIdAndCentreForCentreEmailRegistrationConfirmationHashPair(
+                    DefaultEmail,
+                    DefaultCode
+                )
+            ).Returns((DefaultUserId, DefaultCentreId, DefaultCentreName));
+
+            A.CallTo(
+                () => claimAccountService.GetAccountDetailsForCompletingRegistration(
+                    DefaultUserId,
+                    DefaultCentreId,
+                    DefaultCentreName,
+                    DefaultEmail,
+                    DefaultLoggedInUserId
+                )
+            ).Returns(expectedModel);
+
+            A.CallTo(
+                () => userService.GetUserById(DefaultLoggedInUserId)
+            ).Returns(userEntity);
+
+            // When
+            var result = controllerWithLoggedInUser.LinkDlsAccount(DefaultEmail, DefaultCode);
+
+            // Then
+            A.CallTo(
+                () => userService.GetUserById(DefaultLoggedInUserId)
+            ).MustHaveHappenedOnceExactly();
+
+            result.Should().BeRedirectToActionResult()
+                .WithActionName("AccountAlreadyExists");
+        }
+
+        [Test]
+        public void
+            LinkDlsAccountGet_when_the_claim_email_address_matches_the_primary_email_of_another_user_redirects_to_WrongUser()

```

Comment:
You're testing the validator method multiple times in the same way, just via different endpoints.
I wouldn't recommend doing this now, but in the future, I _think_ it'd actually be best to pull the validation method out into a Validator class.
Benefits:
* Allows you to do dependency inversion with validator.
* Allows you to mock the validator method rather than mocking the things the method depends on - generally easier.

Reply:
We've talked about this. For now, I've settled on not doing it